### DISCUSSION TITLE
Why reset is called only before 2+ boot and not after each one?

### DISCUSSION THREAD

**Discussion by vudaltsov** [symfony/symfony#48928] [Q&A]
https://github.com/symfony/symfony/blob/9cb6274a5bcfddea10f84369acfe7f480a9c7d1f/src/Symfony/Component/HttpKernel/Kernel.php#L74
https://github.com/symfony/symfony/blob/9cb6274a5bcfddea10f84369acfe7f480a9c7d1f/src/Symfony/Component/HttpKernel/Kernel.php#L111-L125
https://github.com/symfony/symfony/blob/9cb6274a5bcfddea10f84369acfe7f480a9c7d1f/src/Symfony/Component/HttpKernel/Kernel.php#L187-L206

As far as I can understand from the snippets above, resettable services are reseted only before second and ongoing boots.

This behavior is not correct for Monolog BufferHandler, for example. When I get the first request and it buffers some logs, they are not flushed until the next request is handled by this worker.

Am I missing something?

**Comment by mbabker** [comment]
The first time the kernel is booted in a process, it's expected that everything is in a clean state, so there shouldn't be anything that needs to be reset on the first boot.

**Reply by vudaltsov** [reply]
That's alright, but why not cleaning at the end of each request?

**Reply by mbabker** [reply]
In a traditional process, the kernel only handles a single request.  Resetting the kernel state just to shut everything down is going to be a waste of resources in an overwhelming majority of cases.

The buffer handler is a bit of a special thing, it won't flush its buffer until either its `reset()` method is explicitly called (which will happen when resetting the kernel) or the PHP process shuts down (as it overloads `__destruct()` and has a `register_shutdown_function()` call inside it).  So for a long-running process where it's handling multiple requests, it's one of those things where the behavior that works well for the majority of users hits a bit of a snag.  I don't use any of those environments for a web facing frontend, so I'm not sure what the best way to deal with that is myself, but for one of our daemon CLI processes, we do use https://github.com/LongRunning/LongRunning along with a custom cleaner to reset our loggers on a scheduled interval and it might work for your use case.
